keps/sig-api-machinery/4420-retry-generate-name/README.md

```diff
@@ -0,0 +1,541 @@
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- Verify that retries are not attempted when feature is disabled
+- Verify that up to max retries are attempted when conflicting names exist
+- Verify that if max retries is exceeded that a already exists error is returned
+
+##### e2e tests
+
+This feature is not e2e testable. There is no predictable way to
+trigger name collisions.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Unit tests added. The `generic/registry/store_test.go` suite covers this functionality remarkably well.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Add integration tests
+- Additional tests are in Testgrid and linked in KEP
+- Benchmark worst case: Max retries are attempted for all create requests.
+
+#### GA
+
+- Add `generate_name_retries: {count}` label to `apiserver_request_duration_seconds`
+  metric.
+
+### Upgrade / Downgrade Strategy
+
+N/A. This only reduces the probability of name conflicts when a request is
+served by an apiserver with this feature enabled.
+
+### Version Skew Strategy
+
+N/A. This only reduces the probability of name conflicts when a request is
+served by an apiserver with this feature enabled.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: RetryGenerateName
+  - Components depending on the feature gate: kube-apiserver
+
+
+###### Does enabling the feature change any default behavior?
+
+No. It only reduces the probability of a error that can occur randomly.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+This is safe. The feature becomes enabled again.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+N/A
+
+###### What specific metrics should inform a rollback?
+
+`apiserver_request_duration_seconds` for create requests.  If this metric
+increases significantly when this feature is enabled, it may indicate
+an issue with this feature.
+
+We also plan to add a `generate_name_retries: {count}` label to this metric
+to make it easy to inspect the latency of create requests where retries
+occurred.
```

Comment:
Would there be a way for someone to determine what proportion of `generateName` **create** requests needed at least one retry? (I don't think this metric would provide that).

Reply:
> Would there be a way for someone to determine what proportion of `generateName` **create** requests needed at least one retry? (I don't think this metric would provide that).

What would we suggest a cluster-admin do with the information that varies from "if you see a lot of retries, the feature may be having a problem"?

Reply:
Almost certainly the feature has a problem. There is also a small possibility that millions of names were generated with the same prefix and collisions have becoming common.. this requires an exotic etcd setup or an alternate storage backend, but it is possible..  I'll clarify.

Should I simplify the metric to just track if retries happened, and not track the count?